stays/views.py

```diff
@@ -1,3 +1,52 @@
-from django.shortcuts import render
+from datetime import datetime
+
+from django.http      import JsonResponse
+from django.views     import View
+from django.db.models import Q, F, Count, Sum, Min, Case, When, Subquery, OuterRef
+
+from stays.models import Staytype, RoomOption
+
+class StaytypeListView(View):
+    def get(self, request):
+        category_id = request.GET.get('category', None)
+        location    = request.GET.get('location', None)
+        check_in    = request.GET.get('CheckIn', None)
+        check_out   = request.GET.get('CheckOut', None)
+
+        check_in  = datetime.strptime(check_in, '%Y-%m-%d')
+        check_out = datetime.strptime(check_out, '%Y-%m-%d')
+
+        q = Q()
+
+        if category_id:
+            q &= Q(category_id = category_id) 
+
+        if location:
+            q &= Q(address__icontains = location) 
+
+        stays = Staytype.objects.filter(q)
+
+        if check_in and check_out:
+            q1 = Q(room__book__check_in__lte = check_in, room__book__check_out__gte = check_in)
+            q2 = Q(room__book__check_in__lte = check_out, room__book__check_out__gte = check_out)
+            q3 = Q(room__book__check_in__lte = check_in, room__book__check_out__gte = check_out)
+
+            stays = Staytype.objects.filter(q)
+
+            total = stays.annotate(total = Sum('room__quantity')).filter(pk = OuterRef('pk'))
+            booked = stays.annotate(booked = Count('room__book__id',filter=(q1|q2|q3))).filter(pk = OuterRef('pk'))
+            stays = stays.annotate(total = Subquery(total.values('total')), booked = Subquery(booked.values('booked')),remain = F('total')-F('booked'), \
+                is_available = Case(When(remain = 0), then = False, default = True)).exclude(is_available= False)
```

Comment:
이 문제를 해결하시다니 정말 👍👍👍
이번 코드는 문제해결 했던 방법 꼭 블로깅 하셔서 취업하실때 사용하시면 너무나도 좋을 것 같네요~

지금 코드도 너무 좋지만 피드백을 드리자면, 아래와 같은 방향으로 `line34~39` 코드를 수정해보시면 어떨까요?

sql 쿼리문 자체도 좀 더 간결해지고 가독성도 좋아질 것 같아요~ (지금까지 작성하신 것도 사실 대단하기 때문에, 기능 구현이 많이 남았으면 다음으로 수정을 미루셔도 됩니다! 코멘트 남겨주세요~)

```suggestion
        stays = Staytype.objects.annotate(
            total_room_count  = Sum('room__quantity'),
            booked_room_count = Subquery(
                Staytype.objects.annotate(booked_count = Count('room__book__id', filter=(q1| q2| q3))).filter(pk = OuterRef('pk')).values('booked_count')
            ),
            is_available = Case(
                When(total_room_count__gte = F('booked_room_count'), then=True), 
                default = False, 
                output_field = IntegerField()
            )
        ).exclude(is_available= False)
```

Reply:
병민님과 같은 팀원 민님의 도움으로 만들 수 있었습니다. 정말 감사드려요!!! 리뷰 토대로 리팩토링 했습니다.